### DISCUSSION TITLE
Got different DirectoryInfo.FullName for same target

### DISCUSSION THREAD

**Discussion by hsupu** [PowerShell/PowerShell#18457] [General]
It may or may not a pwsh issue - dotnet is suspecious too.

```ps1
$a = Get-Item "C:\dir"
$b = Get-Item "C:\dir\"

if ($a.FullName -eq $b.FullName) {
    Write-Host "expected"
}
```

It shows that DirectoryInfo keeps the trailing slash somehow, which makes this inconsistent.

Tested on pwsh 7.2.7 + Windows 11.

**Comment by rhubarb-geek-nz** [comment]
In general you can't compare two files for equivalence just by comparing the filenames. For example you could have different case for the drive letter or path, likewise if there is a symbolic or hard-link in the path, or if it is a network drive and comparing a mounted drive versus a UNC path. When it comes to case, the case sensitivity is a property of the file system itself not the running platform.

**Comment by hsupu** [comment]
Thanks for your reminder, let me clarify my report: It's different because the FullName of one is `C:\dir` and another is `C:\dir\`.

It's not a case about case sensitivity, but I agree that it may relates to symlink / filesystem. I think it maybe a bug so..

**Comment by rhubarb-geek-nz** [comment]
I am not suggesting it is a bug, it is simply a feature of file paths and file names. File paths identity the file but there is no contract that says any other path can't resolve to the same file.

**Comment by hsupu** [comment]
Is there a way to get a normalized path for existing dir? I tried `Resolve-Path (Get-Item "C:\dir" "C:\dir\")` but the results were different too.

**Comment by rhubarb-geek-nz** [comment]
You may have to examine each directory component in the path to get the correct case for the file, eg if you had said "c:\program files" then you would want "C:\Program Files" back again.
Likewise, you would need the short name, "c:\PROGRA~1" translated back to "C:\Program Files".
If you have a network drive "\\\\MYSERVER\MYDIR\MYFILE" do you need that to be equivalent to a network share mapping "M:\MYFILE" or should the "M:\MYFILE" become ""\\\\MYSERVER\MYDIR\MYFILE".

**Comment by rhubarb-geek-nz** [comment]
And if you have a symbolic link along the path, do you need that section replaced by the value of the symbolic link?

**Comment by rhubarb-geek-nz** [comment]
Depending on your actual use-case, it may be simplest to simply remove all trailing slashes ( [System.IO.Path]::DirectorySeparatorChar ) on file paths. However for comparison you need to know whether to do case-sensitive or case-insensitive compare. But you can't remove the trailing slash if it refers to the root directory otherwise you convert the path to a relative directory.

An alternative might be to always add a trailing slash to a directory name if it is not there, that way you can tell if it refers to a file or a directory, and you avoid the problem with root directories being treated differently.

**Comment by hsupu** [comment]
I found a same issue report #4299 . For my case, I verify each folder and add trailing slash then compare them.
